File: algorithms/1_easy/A/00067_add_binary/solution.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:

        a = list(a)
        b = list(b)

        carry = 0
        ord_0 = ord('0')

        if len(a) < len(b):
            a, b = b, a

        for i in range(len(a) - 1, -1, -1):

            b_char = '0'

            if b:
                b_char = b.pop()

            bit_sum = (ord(a[i]) - ord_0) + (ord(b_char) - ord_0) + carry
            carry = 0
            if bit_sum > 1:
                bit_sum -= 2
                carry = 1

            a[i] = chr(bit_sum + ord_0)

            if carry == 0 and not b:
                break

        if carry == 1:
            a.insert(0, '1')

        return "".join(a)

File: algorithms/1_easy/A/00067_add_binary/solution.rs
impl Solution {
    pub fn add_binary(a: String, b: String) -> String {

        let mut carry = 0_u8;
        let mut a_bytes = a.into_bytes();
        let mut b_bytes = b.into_bytes();

        if a_bytes.len() < b_bytes.len() {
            std::mem::swap(&mut a_bytes, &mut b_bytes);
        }

        let a_bytes_len = a_bytes.len();

        for i in (0..a_bytes_len).rev() {

            let mut b_byte = b'0';

            if let Some(byte) = b_bytes.pop() {
                b_byte = byte;
            }

            let mut bit_sum = (a_bytes[i] - b'0') + (b_byte - b'0') + carry;
            carry = 0;
            if bit_sum > 1 {
                bit_sum -= 2;
                carry = 1;
            }

            a_bytes[i] = bit_sum + b'0';

            if carry == 0 && b_bytes.is_empty() {
                break;
            }
        }

        if carry == 1 {
            a_bytes.insert(0_usize, b'1');
        }

        String::from_utf8(a_bytes).unwrap()
    }
}
